File: nikkake_flutter/lib/domain/stats.dart
import '../models/models.dart';
import 'date_utils.dart';

/// 進捗画面の集計。すべて純粋関数。

/// その日「やった」と数える基準。
///
/// 全セット完了(completed)だけを対象にすると、4種目のうち3種目やった日が
/// ノーカウントになりストリークが切れる。習慣化アプリとしてそれは逆効果なので、
/// 中断(skipped)以外はすべて実施した日として扱う。
bool didWorkout(RoutineLog log) => log.status != LogStatus.skipped;

/// 連続日数。
/// 「今日まだやっていない」だけで途切れた表示になるのは体験が悪いので、
/// 最終実施日が今日か昨日ならストリークは生きているものとして数える。
StreakInfo calculateStreak(List<RoutineLog> logs, [DateTime? now]) {
  final today = now ?? DateTime.now();
  final dates = logs.where(didWorkout).map((l) => l.logDate).toSet().toList()..sort();

  if (dates.isEmpty) {
    return const StreakInfo(current: 0, longest: 0);
  }

  var longest = 1;
  var run = 1;
  for (var i = 1; i < dates.length; i++) {
    final gap = daysBetween(parseDateString(dates[i - 1]), parseDateString(dates[i]));
    run = gap == 1 ? run + 1 : 1;
    if (run > longest) longest = run;
  }

  final lastDate = dates.last;
  final gapFromToday = daysBetween(parseDateString(lastDate), today);

  var current = 0;
  if (gapFromToday <= 1) {
    current = 1;
    for (var i = dates.length - 1; i > 0; i--) {
      final gap = daysBetween(parseDateString(dates[i - 1]), parseDateString(dates[i]));
      if (gap != 1) break;
      current++;
    }
  }

  return StreakInfo(current: current, longest: longest, lastCompletedDate: lastDate);
}

/// 直近n日分の実施状況。記録の無い日も0で埋めて返す
List<DailyStats> calculateDailyStats(List<RoutineLog> logs, int days, [DateTime? now]) {
  final today = now ?? DateTime.now();
  final byDate = <String, List<RoutineLog>>{};
  for (final log in logs) {
    byDate.putIfAbsent(log.logDate, () => []).add(log);
  }

  final result = <DailyStats>[];
  for (var i = days - 1; i >= 0; i--) {
    final date = getDateString(addDays(today, -i));
    final dayLogs = byDate[date] ?? const <RoutineLog>[];
    result.add(DailyStats(
      date: date,
      completedCount: dayLogs.where(didWorkout).length,
      totalCount: dayLogs.length,
    ));
  }
  return result;
}

/// カレンダー用の「実施した日」の集合
Set<String> completedDateSet(List<RoutineLog> logs) =>
    logs.where(didWorkout).map((l) => l.logDate).toSet();

/// 種目ごとの推移。重量種目は最大重量、時間・回数種目は合計が伸びを表す
List<ExerciseProgressPoint> calculateExerciseProgress(
  String exerciseId,
  List<RoutineLog> routineLogs,
  List<ExerciseLog> exerciseLogs,
) {
  final dateByLogId = {for (final l in routineLogs) l.id: l.logDate};
  final byDate = <String, List<ExerciseLog>>{};

  for (final log in exerciseLogs) {
    if (log.exerciseId != exerciseId) continue;
    final date = dateByLogId[log.routineLogId];
    if (date == null) continue;
    byDate.putIfAbsent(date, () => []).add(log);
  }

  final points = byDate.entries.map((entry) {
    var maxWeight = 0.0;
    var totalReps = 0;
    var totalVolume = 0.0;

    for (final s in entry.value) {
      final w = s.actualWeight ?? 0;
      final r = s.actualReps ?? 0;
      if (w > maxWeight) maxWeight = w;
      totalReps += r;
      totalVolume += w * r;
    }

    return ExerciseProgressPoint(
      date: entry.key,
      maxWeight: maxWeight,
      totalReps: totalReps,
      totalVolume: totalVolume,
    );
  }).toList()
    ..sort((a, b) => a.date.compareTo(b.date));

  return points;
}

/// そのワークアウトの総挙上量(kg×回)。自重種目は0になる
double calculateVolume(List<ExerciseLog> logs) => logs.fold(
      0.0,
      (sum, l) => sum + (l.actualWeight ?? 0) * (l.actualReps ?? 0),
    );

OverallStats calculateOverallStats(
  List<RoutineLog> routineLogs,
  List<ExerciseLog> exerciseLogs, [
  DateTime? now,
]) {
  final today = now ?? DateTime.now();
  final done = routineLogs.where(didWorkout).toList();
  final weekAgo = getDateString(addDays(today, -6));

  return OverallStats(
    totalWorkouts: done.length,
    totalDurationSec: done.fold(0, (sum, l) => sum + (l.durationSec ?? 0)),
    totalSets: exerciseLogs.length,
    thisWeekCount: done.where((l) => l.logDate.compareTo(weekAgo) >= 0).length,
  );
}

File: nikkake_react_native/src/lib/stats.ts
import { DailyStats, ExerciseLog, ExerciseProgress, RoutineLog, StreakInfo } from '../../types';
import { addDays, daysBetween, getDateString, parseDateString } from './utils';

/**
 * 進捗画面の集計。すべて純粋関数にしてあるので、ストレージを触らずに単体テストできる。
 */

/**
 * その日「やった」と数える基準。
 *
 * 全セット完了(completed)だけを対象にすると、4種目のうち3種目やった日が
 * ノーカウントになりストリークが切れる。習慣化アプリとしてそれは逆効果なので、
 * 中断(skipped)以外はすべて実施した日として扱う。
 */
export const didWorkout = (log: RoutineLog): boolean => log.status !== 'skipped';

/**
 * 連続日数。
 * 「今日まだやっていない」だけでストリークが途切れた表示になるのは体験が悪いので、
 * 最終実施日が今日か昨日ならストリークは生きているものとして数える。
 */
export const calculateStreak = (logs: RoutineLog[], today: Date = new Date()): StreakInfo => {
  const completedDates = Array.from(new Set(logs.filter(didWorkout).map(l => l.log_date))).sort();

  if (completedDates.length === 0) {
    return { current: 0, longest: 0, lastCompletedDate: null };
  }

  // 最長ストリーク
  let longest = 1;
  let run = 1;
  for (let i = 1; i < completedDates.length; i++) {
    const gap = daysBetween(parseDateString(completedDates[i - 1]), parseDateString(completedDates[i]));
    run = gap === 1 ? run + 1 : 1;
    longest = Math.max(longest, run);
  }

  // 現在のストリーク
  const lastCompletedDate = completedDates[completedDates.length - 1];
  const gapFromToday = daysBetween(parseDateString(lastCompletedDate), today);

  let current = 0;
  if (gapFromToday <= 1) {
    current = 1;
    for (let i = completedDates.length - 1; i > 0; i--) {
      const gap = daysBetween(parseDateString(completedDates[i - 1]), parseDateString(completedDates[i]));
      if (gap !== 1) break;
      current++;
    }
  }

  return { current, longest, lastCompletedDate };
};

/** 直近n日分の実施状況。カレンダーと棒グラフの元データ */
export const calculateDailyStats = (
  logs: RoutineLog[],
  days: number,
  today: Date = new Date()
): DailyStats[] => {
  const byDate = new Map<string, RoutineLog[]>();
  for (const log of logs) {
    const bucket = byDate.get(log.log_date) ?? [];
    bucket.push(log);
    byDate.set(log.log_date, bucket);
  }

  const result: DailyStats[] = [];
  for (let i = days - 1; i >= 0; i--) {
    const date = getDateString(addDays(today, -i));
    const dayLogs = byDate.get(date) ?? [];
    result.push({
      date,
      completedCount: dayLogs.filter(didWorkout).length,
      totalCount: dayLogs.length,
    });
  }

  return result;
};

/** カレンダー用の「実施した日」の集合 */
export const completedDateSet = (logs: RoutineLog[]): Set<string> =>
  new Set(logs.filter(didWorkout).map(l => l.log_date));

/**
 * 種目ごとの推移。重量種目は最大重量、時間種目は合計時間が伸びを表すので両方返す。
 */
export const calculateExerciseProgress = (
  exerciseId: string,
  routineLogs: RoutineLog[],
  exerciseLogs: ExerciseLog[]
): ExerciseProgress[] => {
  const dateByLogId = new Map(routineLogs.map(l => [l.id, l.log_date]));
  const byDate = new Map<string, ExerciseLog[]>();

  for (const log of exerciseLogs) {
    if (log.exercise_id !== exerciseId) continue;
    const date = dateByLogId.get(log.routine_log_id);
    if (!date) continue;

    const bucket = byDate.get(date) ?? [];
    bucket.push(log);
    byDate.set(date, bucket);
  }

  return Array.from(byDate.entries())
    .map(([date, sets]) => ({
      date,
      maxWeight: sets.reduce((max, s) => Math.max(max, s.actual_weight ?? 0), 0),
      totalReps: sets.reduce((sum, s) => sum + (s.actual_reps ?? 0), 0),
      totalVolume: sets.reduce((sum, s) => sum + (s.actual_weight ?? 0) * (s.actual_reps ?? 0), 0),
    }))
    .sort((a, b) => a.date.localeCompare(b.date));
};

/** そのワークアウトの総挙上量(kg×回)。自重種目は0になる */
export const calculateVolume = (logs: ExerciseLog[]): number =>
  logs.reduce((sum, l) => sum + (l.actual_weight ?? 0) * (l.actual_reps ?? 0), 0);

export interface OverallStats {
  totalWorkouts: number;
  totalDurationSec: number;
  totalSets: number;
  thisWeekCount: number;
}

export const calculateOverallStats = (
  routineLogs: RoutineLog[],
  exerciseLogs: ExerciseLog[],
  today: Date = new Date()
): OverallStats => {
  const completed = routineLogs.filter(l => l.status !== 'skipped');
  const weekAgo = getDateString(addDays(today, -6));

  return {
    totalWorkouts: completed.length,
    totalDurationSec: completed.reduce((sum, l) => sum + (l.duration_sec ?? 0), 0),
    totalSets: exerciseLogs.length,
    thisWeekCount: completed.filter(l => l.log_date >= weekAgo).length,
  };
};
